### DISCUSSION TITLE
Asset Catalog Assignment

### DISCUSSION THREAD

**Discussion by bdsoha** [skyvar/.github#6] [Announcements]
@skyvar/everyone A new assignment has been published for those who have completed the previous one.

Please read through the [requirements](https://github.com/skyvar/asset-catalog) carefully. Any questions regarding the assignment should be posted here, I will make changes to the `README.md` if needed.

**Comment by chanaSlumianski** [comment]
Is minio called Asset Catalog?

**Reply by bdsoha** [reply]
MinIO is an S3 compatible object storage.
It most probably will be part of the *server architecture*, but for now focus on the client.

**Comment by RivkaZishinski** [comment]
- AWS has an easy user interface, which makes it attractive for beginners,
- MinIO is a high performance, distributed object storage system
AWS compatible minIO.
That's why I'm thinking of using MINIO
I would love to know if I am on the right way to understanding the task.

**Reply by bdsoha** [reply]
[As mentioned above](https://github.com/orgs/skyvar/discussions/6#discussioncomment-5835652), MinIO can be part of the server architecture.
For now, focus on the client.

**Comment by RivkaZishinski** [comment]
Server architecture is good with minIO.
In focus on the client side: 
Typer CLI in Python -  is example to CLI - tool?

**Reply by bdsoha** [reply]
Python is one of many options for a language. Assuming Python, Typer is one example of many packages.

**Reply by RivkaZishinski** [reply]
short update:
I opened a Python image and installed a typer on it
And I'm learning the style of working with typer

**Comment by MiriamRotenshtein** [comment]
Is the CLI tool the client?

**Reply by bdsoha** [reply]
Yes.

**Reply by MiriamRotenshtein** [reply]
Should I write only the server side and add a CLI tool?

**Reply by bdsoha** [reply]
Did you read the comments above?

**Reply by MiriamRotenshtein** [reply]
Is `npm start` in cmd for example considered a CLI tool?
Doesn't this mean typer in Python, commander in node.js, etc.?

**Reply by bdsoha** [reply]
CLI stands for *command line interface*, meaning, a program that is executed from the command-line without a *graphical interface*.
Yes, `npm` is an example of a CLI tool, so is `git` *(as well as many others)*.

There is no single answer to which *package* or language to use.
Yes, `Typer` in `Python` is one example, however, there are many more.

**Comment by RivkaBerlinski** [comment]
According to what I understand in this task we will have to manage parallel processes in a distributed manner.
As it seems to me, `RabbitMQ` can be the right tool for managing the processes.
Am I right?

**Reply by bdsoha** [reply]
Nope. `RabbitMQ` is a message-queue and does not manage processes.
It can *(in some cases)* allow distributed process to inner communicate.
Managing process pools is an integrated part of all modern languages, without the need of external tooling.

This task is focused on developing a client, not a server.
Usually, a client is not granted access to a backend message-bus.

**Reply by RivkaBerlinski** [reply]
After searching and investigating I saw that `multiprocessing` it is recommended as a library both for managing processes and for keeping a shared memory.
Am I in the right direction?

**Reply by RivkaBerlinski** [reply]
Do you have an answer for me?

**Reply by bdsoha** [reply]
Choosing a package is an implementation step, I did not have a specific package in-mind when writing this assignment.
The main focus was on **requirements**, when choosing the implementation you must verify that it meets the requirements.
If it does then use it.

**Comment by chanaSlumianski** [comment]
We tried to start to see how to work with React since it is a client language.
We started a project in React building it with CLI and Node
By the command `npx create-ink-app` and on it we started to build,
We wanted to know if this is the direction or if we should look for another way...

**Reply by bdsoha** [reply]
No. React is a **UI**, not a CLI.
Correct, you are using the CLI to generate the code scaffolding of React, but it is still a UI.

**Reply by bdsoha** [reply]
`node`, on the other hand, can be used as a CLI application.

**Comment by MiriamRotenshtein** [comment]
Is `commander` in `node.js` is good CLI tool?

**Reply by bdsoha** [reply]
Yes, it is a viable option.

**Comment by FridaPshitik** [comment]
Is using `pystac_client` in python a good way
Or is there a better way?

**Reply by bdsoha** [reply]
How and why are you suggestion a *cataloging* tool for the client frontend? 
You were requested to develop a CLI tool to upload *assets* to a server. 
In later steps, the server will handle cataloging the assets.

**Comment by RivkaZishinski** [comment]
After running Typper in Python, I try to understand the task.
the assets could be of any type: audio, images, videos, text-data, binary-data, etc. 
and the user can store them in distributed system.
all the comunication with the distributed system will be in the server.
but what will be in the client?

> The client will watch a directory for file changes and upload the changed files to the remote server.
> The client will not change or move the files from the watched directory at any point.

how I have the files and the name of files in the client if the server not realy?

> The client must be able to recover state from a previous execution.


how The client recover state from a previous execution.
 If  I not have  any server that able to store?

**Reply by bdsoha** [reply]
1. The client is required to upload the assets to a remote server *(all other cataloging tasks fall upon the server)*.
2. The client can store a global *cache* with a list of filenames or hashes of files that were previously uploaded.

**Comment by ShaindyBorzikowsky** [comment]
I simulate a server (currently a separate folder in the project),
I "upload" files to it,
And want to make sure that the next step is actually to create a cache memory that I can use to make sure that I'm not uploading a file with a name that already exists?

**Reply by bdsoha** [reply]
The cache can be done locally by storing either the *names* or *hashes* of the uploaded file.

**Comment by MiriamRotenshtein** [comment]
I switched to developing in _Python_ after doing various searches and seeing that _Python_ has many libraries that can be used for this task.

**Reply by bdsoha** [reply]
IMO, this project can be written in at least 10 different languages.

**Comment by esterlev23** [comment]
Good morning, I'm back today after a two-day break.
I want to start a task.
I saw that the task was self-work.
But anyway I wanted to ask,
Is it possible to work in pairs? I want to apply together with @RivkaZishinski.

**Reply by bdsoha** [reply]
I would prefer you work alone.
You can assign her to review your PR once it is ready.

**Reply by esterlev23** [reply]
After reading all this discussion, there are things I still don't understand.
I'm trying to understand the requirements:

>The client will not change or move the files from the watched directory at any point.
 
 Isn't it a double save of all files on both the server and the client?

**Reply by bdsoha** [reply]
Not at all.

#### Example Use-Case
You have a local photo library that you want to store in the cloud as a backup.
Do you really want your local data to be *changed*, *moved*, or *removed*?

**Comment by EsterWeiss** [comment]
About:

> 2. The client will *watch* a directory for file changes and upload the changed files to the remote server.

Do you mean *watch* every file that was previously uploaded to a remote server?

**Reply by bdsoha** [reply]
Watch every *local* file that was uploaded to a *remote server*.

**Reply by EsterWeiss** [reply]
About:

> 5. The client must be able to recover state from a previous execution.

1. Do all previous versions need to be saved or is it enough to save one version back?
2. Should the interface allow restoring only the previous state or several states back as well?
(The two questions are interdependent).

**Reply by bdsoha** [reply]
Regarding `1`, there are no _versions_. You need to make sure that the client doesn't waste time or resources by trying to re-upload an asset that was already previously uploaded.

Regarding `2`, the answer is similar to that of `1`. You are not storing **hard** `state` or `versions`. Meaning, you don't care when or how an asset was uploaded, just the **fact** that is was uploaded at some point.

**Reply by SaraLitman** [reply]
> Regarding `1`, there are no _versions_. You need to make sure that the client doesn't waste time or resources by trying to re-upload an asset that was already previously uploaded.
> 
> Regarding `2`, the answer is similar to that of `1`. You are not storing **hard** `state` or `versions`. Meaning, you don't care when or how an asset was uploaded, just the **fact** that is was uploaded at some point.

So - there is no handling of update assets - just add a new one?

**Reply by bdsoha** [reply]
Correct.

**Reply by EsterWeiss** [reply]
1.   If so, what is the purpose of watching the changes of properties that have been uploaded in the past?
   Isn't it enough to check if such a property already exists when uploading just save the file name?

2.   And also, what is meant by  *previous execution*?

**Reply by bdsoha** [reply]
Regarding `1`, The concept is similar to that of DropBox or any other syncing tool, except that this is *write-only*, no deletions.
Assume a filename is changed locally, the old asset will not be removed from the remote server and an additional asset will be uploaded with the new name. Same goes for the properties *(when relevant)*.

Regarding `2`, what happens when the local process is stopped or **dies** during execution.
When it is restarted, will it **waste time** attempting to upload duplicates?

**Reply by RivkaBerlinski** [reply]
Every time I start a process I have to make sure that my local cache is the same as the remote server?

**Reply by bdsoha** [reply]
You are not meant to deffer the logic to the server.
Instead, compare your local cache with the local upload directly.

**Reply by RivkaBerlinski** [reply]
> Regarding 2, what happens when the local process is stopped or dies during execution.
When it is restarted, will it waste time attempting to upload duplicates?

This will only occur when the process is stopped or dead,
In such a case, at the beginning of the process, he downloaded to me the names of the files that are on the remote server to restore a previous state.
Or is there a way to save the local upload even when a process dies?

**Reply by bdsoha** [reply]
Read the previous comments as this was already answered.

**Reply by SaraLitman** [reply]
which previous comment?

**Reply by bdsoha** [reply]
More than 1 time, it was stated that the remote server is not for handling the clients logic.

**Comment by MiriamRotenshtein** [comment]
>The cache can be done locally by storing either the names or hashes of the uploaded file.

If it is possible to save the file names locally,
Is this problem solved:
>Regarding 2, what happens when the local process is stopped or dies during execution.
When it is restarted, will it waste time attempting to upload duplicates?

**Reply by bdsoha** [reply]
Yes.

**Comment by chanaSlumianski** [comment]
Do I need to protect the code in a critical section in case two upload the same file name, etc.?

**Reply by bdsoha** [reply]
Yes

**Comment by RivkaZishinski** [comment]
If what is important to me is only the names of the files - for what purpose is needed?
> You can store client-specific data in fixed directories based on Linux conventions

**Reply by bdsoha** [reply]
Where are you expected to store those names, in-memory?

**Reply by RivkaZishinski** [reply]
> The cache can be done locally by storing either the names or hashes of the uploaded file.

What I understand it to make a file cache (maybe json) and there to save the names file

**Comment by chananaimen** [comment]
Good Morning,
I created a file type class that has a file name.
In my main I wrote a function in which I define a variable of file type and on it I run the functions in the class of checking the file's findings and copying it to the server
But I didn't understand what classes I should build to implement oop.
I am now trying to check but so far I have not found an answer

**Comment by SaraLitman** [comment]
@skyvar/everyone , Once the abovementioned has been completed, each member will be assigned another teams PR.
You are to review your colleagues PR and provide a minimum of 10 suggestions.

The suggestions can be in regard to:

- Verification (does the code answer the requirements).
- Styling.
- Clarity (was the code understood to someone without background information).
- Improvements.
- Mistakes.
- Etc.

Please use this opportunity to utilize the tools provided by GitHub as part of the PR.

The instruction applies to anyone who is currently engaged in this task.
